### PR TITLE
Feat extract file

### PR DESCRIPTION
**Pull request by mcanevet**
This does not work yet but as I don't really understand why, I'm still posting this draft to ask for guidance.
Here is the error message I have in Freebox's tasks list (with `FREEBOX_ROOT=HDD` when running tests):
Capture d’écran 2025-02-15 à 14 21 17

here are the logs:
```
2025-02-15T14:17:46.010+0100 [DEBUG] freebox: Extracting the file HDD/VMs/test-6d4f8a11-3b51-4b1d-b6e2-026b7e8dc5f0.raw.xz: tf_provider_addr=registry.terraform.io/hashicorp/freebox tf_rpc=ApplyResourceChange tf_req_id=7f46a0af-9bb4-7baa-1679-3419d9f253ac tf_resource_type=freebox_remote_file
2025-02-15T14:17:46.010+0100 [DEBUG] freebox: Extracting the file to HDD/VMs/test-6d4f8a11-3b51-4b1d-b6e2-026b7e8dc5f0.raw: tf_provider_addr=registry.terraform.io/hashicorp/freebox tf_rpc=ApplyResourceChange tf_req_id=7f46a0af-9bb4-7baa-1679-3419d9f253ac tf_resource_type=freebox_remote_file
```

Looks like `ExtractFile()` does not return an error because I do not have the `"Failed to extract file"` error message in the logs.

@NikolaLohinski any idea of what I'm doing wrong?

### CODE REVIEW

**Comment by NikolaLohinski** [comment]
Hmm it would have probably been a good idea to setup an integration test in free go for this one. I'll see to that bit and it should give us the mission information about this API's behavior.

Maybe it expects the folder to exist in the destination directory ?

In any case, your current implementation does not wait for the task to finish does it ?

Would be good to either watch the events API if these events are available, or poll the Rest API for task id not.

**Comment by mcanevet** [comment]
You're right, I should have write an integration test in free-go first. Maybe the problem is on that component...
I don't think the issue comes from a missing destination directory because I'm trying to extract in the same directory that already exists (VMs).
Indeed, I do not wait for the task to finish, I assumed the process was synchronous, I'll try to figure out how to do that.

**Comment by NikolaLohinski** [comment]
@mcanevet took me a while but I figured out the problem . The dst/src paths need to be absolute on the Freebox, and start with a leading /. 

I fixed that bit in the free-go client with https://github.com/NikolaLohinski/free-go/pull/26 to prefix paths if forgotten as well as a bunch of other unrelated discrepancies in the API, and added an integration test for it

I then bumped the free-go version in the trunk branch in this provider project.

You should be able to continue from there.

Regading my comment on the synchronousity 

> In any case, your current implementation does not wait for the task to finish does it ?

You can take a peek at https://github.com/NikolaLohinski/free-go/blob/main/integration/file_management_test.go#L132 which waits for the task to finish (in ginkgo/gomega context though), but in your case, it should be similar to other place in the provider in regular go logic. Most of it is already implemented thanks to @holyhope :bow: so you should be able to piggy back on it fairly easily : https://github.com/NikolaLohinski/terraform-provider-freebox/blob/main/internal/utils.go#L334

**Comment by mcanevet** [comment]
Thanks a lot. I'll try to find some time to continue this in the next few days

**Comment by holyhope** [comment]
Hello 👋 
That would be a nice feature!

FYI, I did not expose this field in the first implementation because I expect complications about the `Read()` func.
Imo it needs to ensure that the final file (the extracted one) exists and is the right one. The best would be that it can recover as much as possible.
I think that the integration tests you mentioned would help catch these edge cases.

**Comment by mcanevet** [comment]
It starts to work. I'm wondering how I can remove the extracted file at the end of the test in order to remove the commented lines (@NikolaLohinski any hint?).
Also I have to add more tests (with password, with `delete_archive = true`, with `overwrite = true`...).
Also I have to figure out how to deal with @holyhope's remark about "needs to ensure that the final file (the extracted one) exists and is the right one"

File: internal/resource_remote_file.go [hunk 1]
```diff
@@ -423,6 +504,77 @@ func (v *remoteFileResource) Create(ctx context.Context, req resource.CreateRequ
...
+
+	if !model.Extract.IsNull() {
+		var extract *remoteFileExtractModel
+
+		if diags := model.Extract.As(context.Background(), &extract, basetypes.ObjectAsOptions{}); diags.HasError() {
+			resp.Diagnostics.Append(diags...)
+			return
+		}
+
+		if extract.Enable.ValueBool() {
```
Comment (holyhope) [inline internal/resource_remote_file.go:516]: Related to my previous comment, do we need this flag?
Why not checking if `model.DestinationPath` is not null and then extract it to the destination?
I would even recommend to set the `DestinationPath` as required but the entire `Extract` block as optional.
Except if you prefer marking is as `Computed` which means it is possible to programmatically set the value (eg. based on the downloaded file name)?

File: internal/resource_remote_file.go [hunk 2]
```diff
@@ -330,6 +404,13 @@ func (v *remoteFileResource) Schema(ctx context.Context, req resource.SchemaRequ
 					stringplanmodifier.RequiresReplace(),
 				},
 			},
+			"extract": schema.SingleNestedAttribute{
```
Comment (holyhope) [inline internal/resource_remote_file.go:407]: The Update method could check that the hash remains, that the file is already extracted, and simply move the file if the destination changed.
WDYT?

File: internal/resource_remote_file.go [hunk 3]
```diff
@@ -69,17 +72,73 @@ func (o remoteFileModel) AttrTypes() map[string]attr.Type {
...
+	Overwrite       types.Bool   `tfsdk:"overwrite"`
+}
+
+func (o remoteFileExtractModel) ResourceAttributes() map[string]schema.Attribute {
+	return map[string]schema.Attribute{
+		"enable": schema.BoolAttribute{
+			Required:            true,
+			MarkdownDescription: "Enable extraction",
+		},
+		"destination_path": schema.StringAttribute{
```
Comment (holyhope) [inline internal/resource_remote_file.go:95]: I am doubtful about this field.
When using the output of this resource, should we use `resource.destination_path` or `resource.extract.destination_path`, this would depends on the value of `resource.extract.enable` which make it hard to use.
I suggest to extract the file to the already existing `destination_path`, and maybe introduce a `extract.temporary_download_path` which we would be the path where to download the file.
wdyt?

**Review by holyhope — COMMENTED** [review]
Nice work. There is a lot of good ideas. 👍 
I would be happy to contribute if you want.

File: internal/resource_remote_file.go [hunk 4]
```diff
@@ -69,17 +72,73 @@ func (o remoteFileModel) AttrTypes() map[string]attr.Type {
...
+	DeleteArchive   types.Bool   `tfsdk:"delete_archive"`
+	Overwrite       types.Bool   `tfsdk:"overwrite"`
+}
+
+func (o remoteFileExtractModel) ResourceAttributes() map[string]schema.Attribute {
+	return map[string]schema.Attribute{
+		"enable": schema.BoolAttribute{
+			Required:            true,
+			MarkdownDescription: "Enable extraction",
+		},
```
[the remark this answers is not in the document]
Comment (mcanevet) [reply internal/resource_remote_file.go:94]: I was wondering how to initialize defaults. If we set them in the Schema, can't we get rid of `remoteFileExtractModel{}.defaults()`?
Also, I'm not happy with the `enable` attribute. I aso had the idea of extracting when a `destination_path` is set, but it means that we'd have to duplicate the destination path information if we want to extract in the same path than the archive.

File: internal/resource_remote_file.go (hunk 1 above)
Comment (mcanevet) [reply internal/resource_remote_file.go:516]: Agree, but that would mean that we can't default destination path to the archive dir

File: internal/resource_remote_file.go (hunk 2 above)
Comment (mcanevet) [reply internal/resource_remote_file.go:407]: good idea

**Comment by mcanevet** [comment]
@holyhope thanks for helping me on this. Maybe we should clarify some design before going any further in the development:
- what is the purpose of `remoteFileExtractModel{}.defaults()` if we can set them in the Schema?
- should we have a `enable` attribute and default `destination_path` to the archive dir or should we enforce setting `destination_path` (even if it's redundant with the archive path by default) and trigger the extraction based on it?
- what do we do in Read and Update?
I don't know if you are maintainer of this project but I allowed the maintainers to edit this PR. Maybe @NikolaLohinski can give you maintainer privilege as you contributed to a lot of it?

File: internal/resource_remote_file.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (NikolaLohinski) [reply internal/resource_remote_file.go:94]: We should get rid of the enable field, and instead consider that the feature is enabled if the entire block is defined and otherwise disable the feature.

Regading destination_path, a little duplication for explicitness is always a better bet. Otherwise it might be a btit obscure to the end user

File: internal/resource_remote_file.go (hunk 2 above)
Comment (NikolaLohinski) [reply internal/resource_remote_file.go:407]: Agree good idea

File: internal/resource_remote_file.go (hunk 1 above)
Comment (NikolaLohinski) [reply internal/resource_remote_file.go:516]: Same arguments as in https://github.com/NikolaLohinski/terraform-provider-freebox/pull/26/files/fa10a916250273f1a24d3496129bac9c86cbd4f5#r1973053500

**Comment by NikolaLohinski** [comment]
> * Maybe @NikolaLohinski can give you maintainer privilege as you contributed to a lot of it?

I wouldn't mind but the integration tests run on my own freebox with my home runners, so I would need to have a way to restrict that access :sweat_smile:

**Comment by NikolaLohinski** [comment]
> * what is the purpose of `remoteFileExtractModel{}.defaults()` if we can set them in the Schema?
 
 Not 100% sure here, but probably to handle import logic, that doesn't leverage defaults from the schema.
 
> * should we have a `enable` attribute and default `destination_path` to the archive dir or should we enforce setting `destination_path` (even if it's redundant with the archive path by default) and trigger the extraction based on it?

The redundancy gives explicitness IMHO (see my comments on the code)

> * what do we do in Read and Update?

Checking a content hash would probably be the best way to deal with the remote file(s). If the archive is deleted, then we probably need to store some hash of the archive content to be able to check if anything changed to retrigger the process.

But tbh, his might be overkill/overengineering. We might just start with forbidding the deletion of the archive and call it a day, so that the unarchiving is just a utility that's not changing the default behavior of the resource.

File: internal/resource_remote_file.go (hunk 3 above)
Comment (mcanevet) [reply internal/resource_remote_file.go:95]: and set `delete_archive` to `true` by default?

**Comment by mcanevet** [comment]
I was thinking about Read and Update. Actually my use case for the extract attribute is to use the `freebox_remote_file` to retrieve a `raw.xz` image of a disk image of Talos from a http URL, then resize it using `freebox_virtual_disk` then use it in a `freebox_virtual_machine`.

So the content hash of the extracted file will definitely change in my use case and I do not want to this change to be detected and overwrite each time I run terraform.

Also, maybe we could embed the `source_url`, `extract` and `virtual_size` logic in the `freebox_virtual_machine` resource? Something like this?
```hcl
resource "freebox_virtual_machine" "talos" {
  source_url = "https://factory.talos.dev/image/376567988ad370138ad8b2698212367b8edcb69b5fd68c80be1f2ec7d603b4ba/v1.9.3/nocloud-arm64.raw.xz"

  extract = {
    destination_path = "/Freebox/VMs/talos.raw"
  }

  virtual_size = 20 * 1024 * 1024 * 1024

  disk_path = "/Freebox/VMs/talos.raw"
  disk_type = "raw"
  memory    = 8192
  name      = "talos"
  vcpus     = 2
}
```

**Comment by mcanevet** [comment]
I removed the `enable` attribute

File: internal/resource_remote_file_test.go [hunk 5]
```diff
@@ -2,7 +2,9 @@ package internal_test
 
 import (
 	"path"
+	go_path "path"
```
Comment (NikolaLohinski) [inline internal/resource_remote_file_test.go:5]: Why the renaming when it's already imported ?

**Comment by NikolaLohinski** [comment]
> I was thinking about Read and Update. Actually my use case for the extract attribute is to use the `freebox_remote_file` to retrieve a `raw.xz` image of a disk image of Talos from a http URL, then resize it using `freebox_virtual_disk` then use it in a `freebox_virtual_machine`.
> 
> 
> 
> So the content hash of the extracted file will definitely change in my use case and I do not want to this change to be detected and overwrite each time I run terraform.
> 
> 
> 
> Also, maybe we could embed the `source_url`, `extract` and `virtual_size` logic in the `freebox_virtual_machine` resource? Something like this?
> 
> ```hcl
> 
> resource "freebox_virtual_machine" "talos" {
> 
>   source_url = "https://factory.talos.dev/image/376567988ad370138ad8b2698212367b8edcb69b5fd68c80be1f2ec7d603b4ba/v1.9.3/nocloud-arm64.raw.xz"
> 
> 
> 
>   extract = {
> 
>     destination_path = "/Freebox/VMs/talos.raw"
> 
>   }
> 
> 
> 
>   virtual_size = 20 * 1024 * 1024 * 1024
> 
> 
> 
>   disk_path = "/Freebox/VMs/talos.raw"
> 
>   disk_type = "raw"
> 
>   memory    = 8192
> 
>   name      = "talos"
> 
>   vcpus     = 2
> 
> }
> 
> ```

I wouldn't overload the VM resource with extract and other archive related things. It's seem like it will just be hard to maintain.

We should not have the delete_archive parameter whatsoever, as it breaks the logic of the TF resource: the file gets created and deleted as part of the same resource which is an anti-pattern. It only make sense if we track the content of the archive as being the actual resource artefact, meaning that the archive itself is just a temporary artefact.

So my 2 cents for now : remove the delete_archive and let's consider that the remote file resource is the ZIP, and that the extracted part is fire and forget logic.

Hope that makes sense. (I also left you a nitpick comment 😅)

File: internal/resource_remote_file_test.go (hunk 5 above)
Comment (mcanevet) [reply internal/resource_remote_file_test.go:5]: this is weird, this import is automatically added by my IDE...
when I remove it, it automatically add it again

**Comment by mcanevet** [comment]
@NikolaLohinski @holyhope I think we're good now. WDYT?

**Review by NikolaLohinski — APPROVED** [review]

**Comment by NikolaLohinski** [comment]
Closes #19
